test/cql-pytest/test_prepare.py

```diff
@@ -0,0 +1,154 @@
+        yield table
+
+# The following tests test the generation of "pk indexes"
+# "pk indexes" tell the driver which bind variable values it should use to calculate the partition token, so that it can send queries to the correct shard.
+# https://github.com/apache/cassandra/blob/1959502d8b16212479eecb076c89945c3f0f180c/doc/native_protocol_v4.spec#L699-L707
+
+
+# Test generating pk indexes for a single column partition key.
+def test_single_pk_indexes(cql, table1):
+    prepared = cql.prepare(f"SELECT p FROM {table1} WHERE p = ?")
+    assert prepared.routing_key_indexes == [0]
+
+    prepared = cql.prepare(f"SELECT p, c FROM {table1} WHERE c = ? AND p = ?")
+    assert prepared.routing_key_indexes == [1]
+
+# Test that pk indexes aren't generated when the partition key column isn't restricted using a bind variable.
+# In this situation the driver won't be able to calculate the token, so pk indexes should be empty (None).
+def test_single_pk_no_indexes(cql, table1):
+    prepared = cql.prepare(f"SELECT p, c FROM {table1}")
+    assert prepared.routing_key_indexes is None
+
+    prepared = cql.prepare(f"SELECT p, c FROM {table1} WHERE c = ? ALLOW FILTERING")
+    assert prepared.routing_key_indexes is None
+
+    prepared = cql.prepare(f"SELECT p FROM {table1} WHERE p = 0 AND c = ?")
+    assert prepared.routing_key_indexes is None
+
+# Test generating pk indexes for a composite partition key.
+def test_composite_pk_indexes(cql, table2):
+    prepared = cql.prepare(
+        f"SELECT * FROM {table2} WHERE p1 = ? AND p2 = ? AND p3 = ? AND p4 = ? AND c1 = ? AND c2 = ?")
+    assert prepared.routing_key_indexes == [0, 1, 2, 3]
+
+    prepared = cql.prepare(
+        f"SELECT * FROM {table2} WHERE p4 = ? AND p3 = ? AND p2 = ? AND p1 = ? AND c1 = ? AND c2 = ?")
+    assert prepared.routing_key_indexes == [3, 2, 1, 0]
+
+    prepared = cql.prepare(
+        f"SELECT * FROM {table2} WHERE r1 = ? AND c2 = ? AND p3 = ? AND p1 = ? AND r2 = ? AND p4 = ? AND c1 = ? AND p2 = ? ALLOW FILTERING")
+    assert prepared.routing_key_indexes == [3, 7, 2, 5]
+
+# Test that pk indexes aren't generated when not all partition key columns are restricted using bind variables.
+# In this situation the driver won't be able to calculate the token, so pk indexes should be empty (None).
+def test_composite_pk_no_indexes(cql, table2):
+    prepared = cql.prepare(
+        f"SELECT * FROM {table2}")
+    assert prepared.routing_key_indexes is None
+
+    prepared = cql.prepare(
+        f"SELECT * FROM {table2} WHERE p1 = ? ALLOW FILTERING")
+    assert prepared.routing_key_indexes is None
+
+    prepared = cql.prepare(
+        f"SELECT * FROM {table2} WHERE p1 = ? AND p2 = ? AND p4 = ? ALLOW FILTERING")
+    assert prepared.routing_key_indexes is None
+
+    prepared = cql.prepare(
+        f"SELECT * FROM {table2} WHERE p1 = ? AND p2 = 0 AND p3 = ? AND p4 = ?")
+    assert prepared.routing_key_indexes is None
```

Comment:
Interesting... So how does this case work in the token-aware driver? If it doesn't work, is it document somewhere that it doesn't and all partition key columns should be bound?
By the way, if all the partition key columns are constant, not bound, the token-aware driver doesn't work as well?

Reply:
Rust documentation contains a `Warning` about it in the section about Prepared statements:
> Warning
For token/shard aware load balancing to work properly, all partition key values must be sent as bound values (see [performance section](https://rust-driver.docs.scylladb.com/stable/queries/prepared.html#performance))

https://rust-driver.docs.scylladb.com/stable/queries/prepared.html

Yeah shard aware routing works only when all partition key values are sent as bound values, in all other cases the query gets sent to a random node